pkg/spoke/managedcluster/healthcheck_controller.go

```diff
@@ -0,0 +1,90 @@
+	clusterv1listers "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	"github.com/open-cluster-management/registration/pkg/helpers"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	discovery "k8s.io/client-go/discovery"
+)
+
+// managedClusterHealthCheckController checks the kube-apiserver health on managed cluster to determine it whether is available.
+type managedClusterHealthCheckController struct {
+	clusterName      string
+	hubClusterClient clientset.Interface
+	hubClusterLister clusterv1listers.ManagedClusterLister
+	discoveryClient  discovery.DiscoveryInterface
+}
+
+// NewManagedClusterHealthCheckController creates a managed cluster health check controller on managed cluster.
+func NewManagedClusterHealthCheckController(
+	clusterName string,
+	hubClusterClient clientset.Interface,
+	hubClusterInformer clusterv1informer.ManagedClusterInformer,
+	discoveryClient discovery.DiscoveryInterface,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterHealthCheckController{
+		clusterName:      clusterName,
+		hubClusterClient: hubClusterClient,
+		hubClusterLister: hubClusterInformer.Lister(),
+		discoveryClient:  discoveryClient,
+	}
+
+	return factory.New().
+		WithInformers(hubClusterInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(1*time.Minute).
+		ToController("ManagedClusterHealthCheckController", recorder)
+}
+
+// sync updates managed cluster available condition by checking kube-apiserver health on managed cluster.
+func (c *managedClusterHealthCheckController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	cluster, err := c.hubClusterLister.Get(c.clusterName)
+	if err != nil {
+		return fmt.Errorf("unable to get managed cluster %q from hub: %w", c.clusterName, err)
+	}
+
+	// current managed cluster is not accepted, do nothing.
+	acceptedCondition := helpers.FindManagedClusterCondition(cluster.Status.Conditions, clusterv1.ManagedClusterConditionHubAccepted)
+	if !helpers.IsConditionTrue(acceptedCondition) {
+		return nil
+	}
+
+	// check the kube-apiserver health on managed cluster.
+	availableCondition := clusterv1.StatusCondition{
+		Type:    clusterv1.ManagedClusterConditionAvailable,
+		Status:  metav1.ConditionTrue,
+		Reason:  "ManagedClusterAvailable",
+		Message: "Managed cluster is available",
+	}
+
+	healthStatus := 0
+	result := c.discoveryClient.RESTClient().Get().AbsPath("/healthz").Do(ctx).StatusCode(&healthStatus)
+	if healthStatus != http.StatusOK {
+		availableCondition.Status = metav1.ConditionFalse
+		availableCondition.Reason = "ManagedClusterKubeAPIServerUnavailable"
+		availableCondition.Message = fmt.Sprintf("The kube-apiserver is not ok, status: %d, error: %v", healthStatus, result.Error())
```

Comment:
on non-200 results, you want to show the health checks that failed.  That isn't the same as the error, it's the bytes.

Reply:
refactor this logic